src/actions/write/saveObservations.ts

```diff
@@ -46,98 +48,95 @@ export const saveObservations = async (
   { caller, input }: PstAction,
 ): Promise<ContractWriteResult> => {
   // get all other relevant state data
-  const { observations, gateways, settings } = state;
-  const { observerReportTxId, failedGateways } = new SaveObservations(input); // does validation on constructor
-  const currentEpochStartHeight = getEpochStart({
-    startHeight: DEFAULT_START_HEIGHT,
-    epochBlockLength: DEFAULT_EPOCH_BLOCK_LENGTH,
-    height: +SmartWeave.block.height,
+  const { observations, gateways, settings, distributions } = state;
+  const { observerReportTxId, failedGateways } = new SaveObservations(input);
+  const { epochStartHeight, epochEndHeight } = getEpochBoundariesForHeight({
+    currentBlockHeight: new BlockHeight(+SmartWeave.block.height), // a block height in the middle of the first epoch
+    epochZeroBlockHeight: new BlockHeight(distributions.epochZeroBlockHeight),
+    epochBlockLength: new BlockHeight(DEFAULT_EPOCH_BLOCK_LENGTH),
   });
 
   // get the gateway that is creating the observation
-  const observingGatewayArray = Object.entries(gateways).find(
-    ([gatewayAddress, gateway]: [WalletAddress, Gateway]) =>
-      gatewayAddress === caller || gateway.observerWallet === caller,
+  const observingGateway = Object.values(gateways).find(
+    (gateway: Gateway) => gateway.observerWallet === caller,
```

Comment:
note: this means we only look at observer wallets to find observers, not gateway wallets.

Reply:
i am going to change this, we should find the gatewa that matches the caller - then use the observer wallet from that.

Reply:
updated here - [7b20b45](https://github.com/ar-io/testnet-contract/pull/125/commits/7b20b4538733750410c39577e9c0711be6c7ccd3)

the `gatewayAddress` is optional, if an observer intends to submit a report on behalf of a specific gateway. otherwise it will default the gateway to the caller, and if that address does not exist on the registry (or the assigned observer wallet for that address does not also equal the caller) throws an error.